src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs

```diff
@@ -1104,44 +1157,129 @@ protected virtual void ReportUnassigned(Symbol symbol, SyntaxNode node, int slot
             _alreadyReported[slot] = true;
+
+            return;
+
+            void addDiagnosticForStructThis(Symbol thisParameter, int thisSlot)
+            {
+                Debug.Assert(CurrentSymbol is MethodSymbol { MethodKind: MethodKind.Constructor, ContainingType.TypeKind: TypeKind.Struct });
+                if (TrackImplicitlyInitializedFields)
+                {
+#if DEBUG
+                    bool foundUnassignedField = false;
+#endif
+                    foreach (var field in _emptyStructTypeCache.GetStructInstanceFields(thisParameter.ContainingType))
+                    {
+                        if (_emptyStructTypeCache.IsEmptyStructType(field.Type))
+                            continue;
+
+                        if (field is TupleErrorFieldSymbol)
+                            continue;
+
+                        int slot = VariableSlot(field, thisSlot);
+                        if (slot == -1 || !State.IsAssigned(slot))
+                        {
+                            AddImplicitlyInitializedField(field);
+#if DEBUG
+                            foundUnassignedField = true;
+#endif
+                        }
+                    }
+
+#if DEBUG
+                    Debug.Assert(foundUnassignedField);
+#endif
+                }
+
+                if (compilation.IsFeatureEnabled(MessageID.IDS_FeatureAutoDefaultStructs))
+                {
+                    Diagnostics.Add(ErrorCode.WRN_UseDefViolationThisSupportedVersion, node.Location, thisParameter.Name);
+                }
+                else
+                {
+                    Diagnostics.Add(
+                        ErrorCode.ERR_UseDefViolationThisUnsupportedVersion,
+                        node.Location,
+                        thisParameter.Name,
+                        new CSharpRequiredLanguageVersion(MessageID.IDS_FeatureAutoDefaultStructs.RequiredVersion()));
+                }
+            }
+
+            void addDiagnosticForStructField(int fieldSlot, FieldSymbol fieldSymbol)
+            {
+                var associatedSymbol = fieldSymbol.AssociatedSymbol;
+                var hasAssociatedProperty = associatedSymbol?.Kind == SymbolKind.Property;
+                var symbolName = associatedSymbol?.Kind == SymbolKind.Property ? associatedSymbol.Name : fieldSymbol.Name;
+                if (CurrentSymbol is not MethodSymbol { MethodKind: MethodKind.Constructor, ContainingType.TypeKind: TypeKind.Struct })
+                {
+                    Diagnostics.Add(hasAssociatedProperty ? ErrorCode.ERR_UseDefViolationProperty : ErrorCode.ERR_UseDefViolationField, node.Location, symbolName);
+                    return;
+                }
+
+                var thisSlot = GetOrCreateSlot(CurrentSymbol.EnclosingThisSymbol());
+                while (true)
+                {
+                    var containingSlot = variableBySlot[fieldSlot].ContainingSlot;
+                    if (containingSlot == fieldSlot)
```

Comment:
>containingSlot == fieldSlot [](http://example.com/codeflow?start=24&length=27)

Minor: This seems surprising to check that `fieldSlot` is the same as `containingSlot`. Could we simply check `if (fieldSlot == 0)` instead, and before using `variableBySlot[fieldSlot].ContainingSlot`?